File: Example/SnapshotTests/BPKIconSnapshotTest.m
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import <Backpack/Color.h>
#import <Backpack/Icon.h>
#import <Backpack/Spacing.h>
#import <FBSnapshotTestCase/FBSnapshotTestCase.h>
#import "BPKIconsTestsUtils.h"

typedef UIImageView* (^BPKIconMakeSmallIcon)(BPKSmallIconName);
typedef UIImageView* (^BPKIconMakeLargeIcon)(BPKLargeIconName);
typedef UIImageView* (^BPKIconGenericMakeIcon)(id);


@interface BPKIconSnapshotTest : FBSnapshotTestCase

@end

@implementation BPKIconSnapshotTest

- (void)setUp {
    [super setUp];

    self.recordMode = NO;
}

- (void)testSmallIcon {
    UIImage *icon = [BPKIcon smallIconNamed:@"flight" color:[BPKColor skyBlue]];
    UIImageView *view = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, icon.size.width, icon.size.height)];
    view.image = icon;

    FBSnapshotVerifyView(view, nil)
}

- (void)testLargeIcon {
    UIImage *icon = [BPKIcon largeIconNamed:@"hotels" color:[BPKColor skyBlue]];
    UIImageView *view = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, icon.size.width, icon.size.height)];
    view.image = icon;

    FBSnapshotVerifyView(view, nil)
}

- (void)testAllSmallIcons {
    NSArray<BPKSmallIconName> *allSmallIcons = [[self class] allSmallIconNames];

    UIView *view = [[self class] generateSmallIconGridWithIcons:allSmallIcons makeIconView:^UIImageView *(BPKSmallIconName name) {
        UIImage *icon = [BPKIcon smallTemplateIconNamed:name];
        UIImageView *imageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, icon.size.width, icon.size.height)];
        imageView.image = icon;

        return imageView;
    }];

    FBSnapshotVerifyView(view, nil);
}

- (void)testAllLargeIcons {
    NSArray<BPKLargeIconName> *allLargeIcons = [[self class] allLargeIconNames];

    UIView *view = [[self class] generateLargeIconGridWithIcons:allLargeIcons makeIconView:^UIImageView *(BPKLargeIconName name) {
        UIImage *icon = [BPKIcon largeTemplateIconNamed:name];
        UIImageView *imageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, icon.size.width, icon.size.height)];
        imageView.image = icon;

        return imageView;
    }];

    FBSnapshotVerifyView(view, nil);
}

+ (NSArray<BPKSmallIconName> *)allSmallIconNames {
    NSArray *allIcons = [BPKIconsTestsUtils getAllIcons];
    NSMutableArray *result = [NSMutableArray new];

    for (NSString *icon in allIcons) {
        if ([icon hasSuffix:@"-sm"]) {
            BPKSmallIconName name = [icon substringToIndex:icon.length - 3];
            [result addObject:name];
        }
    }

    return [result copy];
}

+ (NSArray<BPKLargeIconName> *)allLargeIconNames {
    NSArray *allIcons = [BPKIconsTestsUtils getAllIcons];
    NSMutableArray *result = [NSMutableArray new];

    for (NSString *icon in allIcons) {
        if (![icon hasSuffix:@"-sm"]) {
            [result addObject:(BPKLargeIconName) icon];
        }
    }

    return [result copy];
}

+ (UIView *)generateSmallIconGridWithIcons:(NSArray<BPKSmallIconName> *)icons makeIconView:(BPKIconMakeSmallIcon)makeSmallIcon {
    return [self generateIconGridWithIcons:icons
                              makeIconView:makeSmallIcon
                                  iconSize:BPKIcon.concreteSizeForSmallIcon];
}

+ (UIView *)generateLargeIconGridWithIcons:(NSArray<BPKLargeIconName> *)icons makeIconView:(BPKIconMakeLargeIcon)makeLargeIcon {
    return [self generateIconGridWithIcons:icons
                              makeIconView:makeLargeIcon
                                  iconSize:BPKIcon.concreteSizeForLargeIcon];
}

+ (UIView *)generateIconGridWithIcons:(NSArray<id> *)icons makeIconView:(BPKIconGenericMakeIcon)makeIcon iconSize:(CGSize)iconSize {
    NSUInteger iconsPerRow = 10;
    CGFloat numberOfRows = ceil((CGFloat)icons.count / (CGFloat)iconsPerRow);
    CGFloat width = iconsPerRow * iconSize.width + (iconsPerRow - 1) * BPKSpacingMd + BPKSpacingMd * 2;
    CGFloat height = numberOfRows * iconSize.height + (numberOfRows - 1) * BPKSpacingMd + BPKSpacingMd * 2;

    UIView *view = [[UIView alloc] initWithFrame:CGRectMake(0, 0, width, height)];

    for (NSUInteger i = 0; i < icons.count; i++) {
        id icon = icons[i];
        NSUInteger row = i / iconsPerRow;
        NSUInteger col = i % iconsPerRow;

        UIImageView *imageView = makeIcon(icon);
        imageView.tintColor = BPKColor.black;

        CGFloat x = BPKSpacingMd + col * (iconSize.width + BPKSpacingMd);
        CGFloat y = BPKSpacingMd + row * (iconSize.height + BPKSpacingMd);
        imageView.frame = CGRectMake(x, y, iconSize.width, iconSize.height);
        [view addSubview:imageView];
    }

    return view;
}

@end

File: Backpack/Tests/SnapshotTests/BPKIconSnapshotTest.swift
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import XCTest
import Backpack
import SnapshotTesting

final class BPKIconSnapshotTest: XCTestCase {
    
    override func setUpWithError() throws {
        try? super.setUpWithError()
        isRecording = false
    }
    
    func testSmallIcon() {
        // Given
        let sut = BPKIcon.makeSmallIcon(name: .flight, color: BPKColor.coreAccentColor)
        let parentView = UIImageView(
            frame: CGRect(
                x: 0,
                y: 0,
                width: sut.size.width,
                height: sut.size.height
            )
        )
        
        parentView.image = sut
        assertSnapshot(parentView)
    }
    
    func testLargeIcon() {
        // Given
        let sut = BPKIcon.makeLargeIcon(name: .hotels, color: BPKColor.coreAccentColor)
        let parentView = UIImageView(
            frame: CGRect(
                x: 0,
                y: 0,
                width: sut.size.width,
                height: sut.size.height
            )
        )
        
        parentView.image = sut
        assertSnapshot(parentView)
    }
    
    func testAllSmallIcons() {
        // Given
        let views: [() -> UIView] = BPKIconsTestsUtils.allSmallIcons.map { icon in
            func makeView() -> UIView {
                let sut = BPKIcon.makeSmallIcon(name: icon, color: BPKColor.coreAccentColor)
                let parentView = UIImageView(
                    frame: CGRect(
                        x: 0,
                        y: 0,
                        width: sut.size.width,
                        height: sut.size.height
                    )
                )
                
                parentView.image = sut
                return parentView
            }
            
            return makeView
        }
        
        // Then
        assertSnapshot(generateGrid(iconSize: BPKIcon.concreteSizeForSmallIcon, makeIconView: views))
    }
    
    func testAllLargeIcons() {
        // Given
        let views: [() -> UIView] = BPKIconsTestsUtils.allLargeIcons.map { icon in
            func makeView() -> UIView {
                let sut = BPKIcon.makeLargeIcon(name: icon, color: BPKColor.coreAccentColor)
                let parentView = UIImageView(
                    frame: CGRect(
                        x: 0,
                        y: 0,
                        width: sut.size.width,
                        height: sut.size.height
                    )
                )
                
                parentView.image = sut
                return parentView
            }
            
            return makeView
        }
        
        // Then
        assertSnapshot(generateGrid(iconSize: BPKIcon.concreteSizeForLargeIcon, makeIconView: views))
    }

    private func generateGrid(iconSize: CGSize, makeIconView: [() -> UIView]) -> UIView {
        let iconsPerRow: CGFloat = 10
        let numberOfRows = ceil(CGFloat(makeIconView.count) / CGFloat(iconsPerRow))
        let width = iconsPerRow * iconSize.width + (iconsPerRow - 1) * BPKSpacingMd + BPKSpacingMd * 2
        let height = numberOfRows * iconSize.height + (numberOfRows - 1) * BPKSpacingMd + BPKSpacingMd * 2
        
        let iconsView = UIView(frame: CGRect(x: 0, y: 0, width: width, height: height))
        
        for (index, view) in makeIconView.enumerated() {
            let row = index / Int(iconsPerRow)
            let col = index % Int(iconsPerRow)
            
            let imageView = view()
            
            let xCoordinate: CGFloat = BPKSpacingMd + CGFloat(col) * (iconSize.width + BPKSpacingMd)
            let yCoordinate: CGFloat = BPKSpacingMd + CGFloat(row) * (iconSize.height + BPKSpacingMd)
            imageView.frame = CGRect(x: xCoordinate, y: yCoordinate, width: iconSize.width, height: iconSize.height)
            
            iconsView.addSubview(imageView)
        }
        
        iconsView.translatesAutoresizingMaskIntoConstraints = false
        
        NSLayoutConstraint.activate([
            iconsView.widthAnchor.constraint(equalToConstant: width),
            iconsView.heightAnchor.constraint(equalToConstant: height)
        ])
        
        return iconsView
    }
}
